File: app/src/main/java/pl/mosenko/sunnypodlaskie/mvp/setting/SettingsPresenterImpl.java
package pl.mosenko.sunnypodlaskie.mvp.setting;

import android.content.Context;
import android.content.SharedPreferences;

import com.hannesdorfmann.mosby3.mvp.MvpBasePresenter;

import pl.mosenko.sunnypodlaskie.R;
import pl.mosenko.sunnypodlaskie.util.PreferenceWeatherUtil;
import pl.mosenko.sunnypodlaskie.util.WeatherDataAlarmSyncUtil;

/**
 * Created by syk on 08.06.17.
 */

public class SettingsPresenterImpl extends MvpBasePresenter<SettingsContract.View> implements SettingsContract.Presenter, SharedPreferences.OnSharedPreferenceChangeListener {
    private SharedPreferences mSharedPreferences;

    @Override
    public void setSharedPreferences(SharedPreferences sharedPreferences) {
        this.mSharedPreferences = sharedPreferences;
    }

    @Override
    public void onCreatePreferences() {
        if (isViewNotNullAttached()) {
            getView().addInitialSyncTimeSummary(mSharedPreferences);
        }
    }

    @Override
    public void onStart() {
        mSharedPreferences.registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onStop() {
        mSharedPreferences.unregisterOnSharedPreferenceChangeListener(this);
    }

    @Override
    public boolean validateNewSyncTime(String syncTime) {
        if (PreferenceWeatherUtil.parseSyncTimeToPartiallyDate(syncTime) == null) {
            if (isViewNotNullAttached()) {
                getView().showMessageBadSyncTimeFormat();
            }
            return false;
        }
        return true;
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (!isViewAttached()) {
            return;
        }

        final Context context = getView().getContext();
        boolean changedSyncTime = key.equals(context.getString(R.string.pref_sync_time_key));

        if (changedSyncTime) {
            getView().addSyncTimePreferenceSummary(sharedPreferences, key);
            WeatherDataAlarmSyncUtil.startAlarm(context);
        }
    }

    private boolean isViewNotNullAttached() {
        return getView() != null && isViewAttached();
    }
}

File: app/src/main/java/pl/mosenko/sunnypodlaskie/mvp/setting/SettingsPresenterImpl.kt
package pl.mosenko.sunnypodlaskie.mvp.setting

import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import com.hannesdorfmann.mosby3.mvp.MvpBasePresenter
import pl.mosenko.sunnypodlaskie.R
import pl.mosenko.sunnypodlaskie.util.PreferenceWeatherUtil
import pl.mosenko.sunnypodlaskie.util.WeatherDataAlarmSyncUtil

/**
 * Created by syk on 08.06.17.
 */
class SettingsPresenterImpl : MvpBasePresenter<SettingsContract.View>(), SettingsContract.Presenter, OnSharedPreferenceChangeListener {

    private lateinit var sharedPreferences: SharedPreferences

    override fun setSharedPreferences(sharedPreferences: SharedPreferences) {
        this.sharedPreferences = sharedPreferences
    }

    override fun onCreatePreferences() {
        if (isViewNotNullAttached()) {
            view.addInitialSyncTimeSummary(sharedPreferences)
        }
    }

    override fun onStart() {
        sharedPreferences.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onStop() {
        sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
    }

    override fun validateNewSyncTime(syncTime: String): Boolean {
        if (PreferenceWeatherUtil.parseSyncTimeToPartiallyDate(syncTime) == null) {
            if (isViewNotNullAttached()) {
                view.showMessageBadSyncTimeFormat()
            }
            return false
        }
        return true
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        if (!isViewAttached) {
            return
        }
        val context = view.getContext()!!
        val changedSyncTime = key == context.getString(R.string.pref_sync_time_key)
        if (changedSyncTime) {
            view.addSyncTimePreferenceSummary(sharedPreferences, key)
            WeatherDataAlarmSyncUtil.startAlarm(context)
        }
    }

    private fun isViewNotNullAttached(): Boolean {
        return view != null && isViewAttached
    }
}
